<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Query VCF — query_vcf_variantannotation • echotabix</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Query VCF — query_vcf_variantannotation" />
<meta property="og:description" content="# Alternative methods

 #### Rsamtools ####
 tab &amp;lt;- Rsamtools::scanTabix(file = vcf_url, gr)
 #### Rtracklayer ####
 tab &amp;lt;- Rsamtools::TabixFile(vcf_url)
 open(tab)
 tab &amp;lt;- rtracklayer::import(tab, which=gr)
 #### seqminer ####
 tab &amp;lt;- seqminer::tabix.read(tabixFile = vcf_url,
                              tabixRange = coord_range)
 header &amp;lt;- VariantAnnotation::scanVcfHeader(file = path)
" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">echotabix</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/echotabix.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/RajLabMSSM/echotabix/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Query VCF</h1>
    <small class="dont-index">Source: <a href='https://github.com/RajLabMSSM/echotabix/blob/master/R/query_vcf_variantannotation.R'><code>R/query_vcf_variantannotation.R</code></a></small>
    <div class="hidden name"><code>query_vcf_variantannotation.Rd</code></div>
    </div>

    <div class="ref-description">
    <p># Alternative methods
<code>
 #### Rsamtools ####
 tab &lt;- Rsamtools::scanTabix(file = vcf_url, gr)
 #### Rtracklayer ####
 tab &lt;- Rsamtools::TabixFile(vcf_url)
 open(tab)
 tab &lt;- rtracklayer::import(tab, which=gr)
 #### seqminer ####
 tab &lt;- seqminer::tabix.read(tabixFile = vcf_url,
                              tabixRange = coord_range)
 header &lt;- VariantAnnotation::scanVcfHeader(file = path)
</code></p>
    </div>

    <pre class="usage"><span class='fu'>query_vcf_variantannotation</span><span class='op'>(</span>
  <span class='va'>vcf_url</span>,
  <span class='va'>dat</span>,
  samples <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/character.html'>character</a></span><span class='op'>(</span><span class='op'>)</span>,
  genome <span class='op'>=</span> <span class='st'>"GRCh37"</span>,
  save_path <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempfile</a></span><span class='op'>(</span>fileext <span class='op'>=</span> <span class='st'>".vcf.bgz"</span><span class='op'>)</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>vcf_url</th>
      <td><p>URL or path to VCF.</p></td>
    </tr>
    <tr>
      <th>dat</th>
      <td><p>SNP-level data table.</p></td>
    </tr>
    <tr>
      <th>samples</th>
      <td><p>Sample names to subset the VCF by before computing LD.</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>A <code>character</code> or <code>Seqinfo</code> object.</p><ul>
<li><p><code>character</code>: Genome identifier as a single string or named 
        character vector. Names of the character vector correspond to 
        chromosome names in the file. This identifier replaces the genome 
        information in the VCF <code>Seqinfo</code> (i.e., <code>seqinfo(vcf)</code>).
        When not provided, <code>genome</code> is taken from the VCF file header.</p></li>
<li><p><code>Seqinfo</code>: When <code>genome</code> is provided as a <code>Seqinfo</code>
        it is propagated to the VCF output. If seqinfo information can be 
        obtained from the file, 
        (i.e., seqinfo(scanVcfHeader(fl)) is not empty), the output
        <code>Seqinfo</code> is a product of merging the two.        If a param (i.e., ScanVcfParam) is used in the call to <code>readVcf</code>, 
        the seqlevels  of the param ranges must be present in <code>genome</code>.</p></li>
</ul><p></p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Print messages.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="source"><a class="anchor" href="#source"></a>Source</h2>

    <p><a href='https://bioconductor.org/packages/devel/bioc/vignettes/TVTB/inst/doc/VcfFilterRules.html'>
VariantAnnotation filtering</a></p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='va'>BST1</span> <span class='op'>&lt;-</span> <span class='fu'>echodata</span><span class='fu'>::</span><span class='va'><a href='https://rdrr.io/pkg/echodata/man/BST1.html'>BST1</a></span>
<span class='va'>dat</span> <span class='op'>&lt;-</span> <span class='va'>BST1</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>50</span><span class='op'>)</span>, <span class='op'>]</span>
<span class='va'>vcf_url</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span>
    <span class='st'>"ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/release/20110521/"</span>,
    <span class='st'>"ALL.chr4.phase1_release_v3.20101123.snps_indels_svs.genotypes.vcf.gz"</span>
<span class='op'>)</span>
<span class='va'>vcf</span> <span class='op'>&lt;-</span> <span class='fu'>query_vcf_variantannotation</span><span class='op'>(</span>
    vcf_url <span class='op'>=</span> <span class='va'>vcf_url</span>,
    dat <span class='op'>=</span> <span class='va'>dat</span>
<span class='op'>)</span>
<span class='op'>}</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Brian Schilder, Jack Humphrey, Towfique Raj.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


